chain/network/src/stats/metrics.rs

```diff
@@ -6,25 +6,84 @@ use near_metrics::{
 };
 use near_network_primitives::types::{PeerType, RoutedMessageBody};
 use once_cell::sync::Lazy;
-use std::collections::HashMap;
 
-static PEER_CONNECTIONS: Lazy<IntGaugeVec> = Lazy::new(|| {
-    near_metrics::try_create_int_gauge_vec(
-        "near_peer_connections",
-        "Number of connected peers",
-        &["peer_type", "encoding"],
-    )
-    .unwrap()
-});
+/// Labels represents a schema of an IntGaugeVec metric.
+pub trait Labels: 'static {
+    /// Names of the gauge vector labels.
+    fn names() -> Vec<&'static str>;
```

Comment:
perhaps

```
const NAMES: &'static[&'static str]
```

?

Reply:
Too bad that 

```
pub trait Labels: 'static {
    const N: usize;
    const NAMES: [&'static str; Self::N];
    fn values(&self) -> [&'static str; Self::N];
}
```

doesn't work yet :(

Reply:
Although, it doesn't look too bad with just an extra parameter: 

```diff
diff --git a/chain/network/src/peer_manager/connected_peers.rs b/chain/network/src/peer_manager/connected_peers.rs
index 0bbc66567..02fbcbc6e 100644
--- a/chain/network/src/peer_manager/connected_peers.rs
+++ b/chain/network/src/peer_manager/connected_peers.rs
@@ -75,7 +75,7 @@ pub(crate) struct ConnectedPeer {
     /// Connection stats
     pub stats: ArcSwap<Stats>,
     /// prometheus gauge point guard.
-    pub _metric_point: metrics::GaugePoint<metrics::Connection>,
+    pub _metric_point: metrics::GaugePoint<2, metrics::Connection>,
 
     /// A helper data structure for limiting reading, reporting stats.
     pub throttle_controller: ThrottleController,
diff --git a/chain/network/src/stats/metrics.rs b/chain/network/src/stats/metrics.rs
index b80552d20..0d44950e5 100644
--- a/chain/network/src/stats/metrics.rs
+++ b/chain/network/src/stats/metrics.rs
@@ -8,30 +8,27 @@ use near_network_primitives::types::{PeerType, RoutedMessageBody};
 use once_cell::sync::Lazy;
 
 /// Labels represents a schema of an IntGaugeVec metric.
-pub trait Labels: 'static {
-    /// Names of the gauge vector labels.
-    fn names() -> Vec<&'static str>;
-    /// Converts self to a list of label values.
-    /// values().len() should be always equal to names().len().
-    fn values(&self) -> Vec<&'static str>;
+pub trait Labels<const N: usize>: 'static {
+    const NAMES: [&'static str; N];
+    fn values(&self) -> [&'static str; N];
 }
 
 /// Type-safe wrapper of IntGaugeVec.
-pub struct Gauge<L: Labels> {
+pub struct Gauge<const N: usize, L: Labels<N>> {
     inner: IntGaugeVec,
     _labels: std::marker::PhantomData<L>,
 }
 
-pub struct GaugePoint<L: Labels> {
-    gauge: &'static Gauge<L>,
+pub struct GaugePoint<const N: usize, L: Labels<N>> {
+    gauge: &'static Gauge<N, L>,
     point: IntGauge,
 }
 
-impl<L: Labels> Gauge<L> {
+impl<const N: usize, L: Labels<N>> Gauge<N, L> {
     /// Constructs a new prometheus Gauge with schema `L`.
     pub fn new(name: &str, help: &str) -> Self {
         Self {
-            inner: near_metrics::try_create_int_gauge_vec(name, help, &L::names()).unwrap(),
+            inner: near_metrics::try_create_int_gauge_vec(name, help, &L::NAMES).unwrap(),
             _labels: std::marker::PhantomData,
         }
     }
@@ -39,7 +36,7 @@ impl<L: Labels> Gauge<L> {
     /// Adds a point represented by `labels` to the gauge.
     /// Returns a guard of the point - when the guard is dropped
     /// the point is removed from the gauge.
-    pub fn new_point(&'static self, labels: &L) -> GaugePoint<L> {
+    pub fn new_point(&'static self, labels: &L) -> GaugePoint<N, L> {
         GaugePoint { gauge: self, point: self.inc(labels) }
     }
 
@@ -50,7 +47,7 @@ impl<L: Labels> Gauge<L> {
     }
 }
 
-impl<L: Labels> GaugePoint<L> {
+impl<const N: usize, L: Labels<N>> GaugePoint<N, L> {
     /// Updates the label values.
     /// It decrements the gauge for the old label value,
     /// and increments gauge for the new label value.
@@ -61,7 +58,7 @@ impl<L: Labels> GaugePoint<L> {
     }
 }
 
-impl<L: Labels> Drop for GaugePoint<L> {
+impl<const N: usize, L: Labels<N>> Drop for GaugePoint<N, L> {
     fn drop(&mut self) {
         self.point.dec();
     }
@@ -72,16 +69,15 @@ pub struct Connection {
     pub encoding: Option<Encoding>,
 }
 
-impl Labels for Connection {
-    fn names() -> Vec<&'static str> {
-        vec!["peer_type", "encoding"]
-    }
-    fn values(&self) -> Vec<&'static str> {
-        vec![self.type_.into(), self.encoding.map(|e| e.into()).unwrap_or("unknown")]
+impl Labels<2> for Connection {
+    const NAMES: [&'static str; 2] = ["peer_type", "encoding"];
+
+    fn values(&self) -> [&'static str; 2] {
+        [self.type_.into(), self.encoding.map(|e| e.into()).unwrap_or("unknown")]
     }
 }
 
-pub static PEER_CONNECTIONS: Lazy<Gauge<Connection>> =
+pub static PEER_CONNECTIONS: Lazy<Gauge<2, Connection>> =
     Lazy::new(|| Gauge::new("near_peer_connections", "Number of connected peers"));
 
 pub(crate) static PEER_CONNECTIONS_TOTAL: Lazy<IntGauge> = Lazy::new(|| {
```

Consider this approach: allocating a `Vec` for values doesn't sit great with me.

Reply:
Yes, I've spent like 2h yesterday trying to work around this limitation :P, because this extra parameter hurts my eyes.